### DISCUSSION TITLE
Why do curl source code use plain old C, not C++?

### DISCUSSION THREAD

**Discussion by ehyoitsdavid** [curl/curl#9836] [Q&A] [answered]
Unfortunately when I look at the curl source code it contains almost of pure C, not C++ would you please explain why?

But C++ supports OOP and is safer than plain C?

**Comment by bagder** [comment, marked as answer]
C++ is *not* safer and is (in my view) mostly a much bigger footgun and lures developers into silly things. You can do OOP in C as well, you just need to do it "manually" and the  way we do it in curl I think is fine enough. I can't see any benefits with using C++ in curl.

We keep curl in C89 for maximum portability. I also recently blogged about our decision [not to use C99 for now](https://daniel.haxx.se/blog/2022/11/17/considering-c99-for-curl/).

**Comment by jzakrzewski** [comment]
I'd say C++ is safer *if* you adhere to certain practises. After years of practise I kinda automatically write C++ code that has no leaks, and is exception-safe (except some special cases, where I screw this up ;) ).

But portability-wise... Nothing beats C. Especially not C++.

**Reply by icing** [reply]
When curl was started C++ was a train wreck. Nowadays, C++ has improved, and so has Curl. 

Using C++ in its best practise form, would be a rewrite of Curl now. There is a queue of languages which offer rewrites. It's hard to choose from...

**Reply by jzakrzewski** [reply]
> When curl was started C++ was a train wreck. Nowadays, C++ has improved, and so has Curl.

You don't need to tell me that - I'm still fighting with a project that started way back then, written by people that treated C++ as "C with classes". Step by step it's getting modernized and the amount of bugs that could never happen with modern techniques is astonishing.

> Using C++ in its best practise form, would be a rewrite of Curl now. There is a queue of languages which offer rewrites. It's hard to choose from...

Yep, I've seen all those suggestions to re-write in X. They have no clue how much work that would be, and how many bugs and breakages would it create. Not to mention probably having to drop some platforms.

**Reply by danielgustafsson** [reply]
> Not to mention probably having to drop some platforms.

By "some" I think it's safe to say that a rewrite would drop all platforms outside of Linux, macOS, Windows and the BSD's, and probably all TLS stacks besides OpenSSL.
